### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by MMhaidari**
# What is added new
- [ ] Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.
before
before_postsControllers#index
before_userContorller#show

after
after_postsControllers#index
after_userController#show

Integration specs
Use Capybara to write integration tests for each view in your project.

Think on the flow of things and group the tests of page in a way that makes sense. For example, one test would check for the things in the page and another test for clicking links on it.

- [ ] User index page:

- I can see the username of all other users.
- I can see the profile picture for each user.
- I can see the number of posts each user has written.
- When I click on a user, I am redirected to that user's show page.

- [ ] user show page:
- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see the user's bio.
- I can see the user's first 3 posts.
- I can see a button that lets me view all of a user's posts.
- When I click a user's post, it redirects me to that post's show page.
- When I click to see all posts, it redirects me to the user's post's index page.

- [ ] User post index page:
- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see a post's title.
- I can see some of the post's body.
- I can see the first comments on a post.
- I can see how many comments a post has.
- I can see how many likes a post has.
- I can see a section for pagination if there are more posts than fit on the view.
- When I click on a post, it redirects me to that post's show page.

- [ ] Post show page:
- I can see the post's title.
- I can see who wrote the post.
- I can see how many comments it has.
- I can see how many likes it has.
- I can see the post body.
- I can see the username of each commentor.
- I can see the comment each commentor left.

### CODE REVIEW

File: spec/features/users_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      expect(page).to have_content(@user1.bio)
+    end
+
+    it 'should show posts_counter of clicked user' do
+      expect(page).to have_content(@user1.posts_counter)
+    end
+
+    it 'should have a link to see all posts' do
+      expect(page).to have_link('See all posts', href: user_posts_path(@user1))
+    end
```
Comment (descholar-ceo) [inline spec/features/users_show_spec.rb:37]: - [ ] We should have tests that assert you can see the user's first 3 posts.
- [ ] We should have tests that redirect to the post's show page when you click on a user's post.
- [ ] We should have tests that redirect to the user's post's index page when you click on a **see all posts**.

File: spec/features/posts_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  it 'should show user name' do
+    expect(page).to have_content(@user1.name)
+  end
+
+  it 'should show posts_counter of clicked user' do
+    expect(page).to have_content(@user1.posts_counter)
+  end
+  it 'should show all posts' do
```
Comment (descholar-ceo) [inline spec/features/posts_index_spec.rb:21]: - [ ] The correct description is `should show post's title`
```suggestion
  it 'should show post's title' do
```

File: spec/features/posts_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    expect(page).to have_button('Pagination')
+  end
+  scenario 'Pagination should be displayed if there are more posts than fit on the view' do
+    10.times do
+      post = Post.new(author: @user1, title: 'Sample Post', text: 'this is a sample post')
+      post.save
+    end
+    visit user_posts_path(@user1)
+    expect(page).to have_button('Pagination')
+  end
```
Comment (descholar-ceo) [inline spec/features/posts_index_spec.rb]: ## On this file we are missing these test cases:
- [ ] Tests that redirect to the post's show page when you click on a post.
- [ ] Tests that assert you can see the user's profile picture.

File: spec/features/posts_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    end
+    it 'should show comments counter' do
+      expect(page).to have_content(@first_post.comments_counter)
+    end
+    it 'should show likes counter' do
+      expect(page).to have_content(@first_post.likes_counter)
+    end
+    it 'should show name of commenter' do
+      expect(page).to have_content(@comment1.user.name)
+    end
```
Comment (descholar-ceo) [inline spec/features/posts_show_spec.rb]: ### From this spec file, we are missing the following test cases:
- [ ] Tests that assert you can see who wrote the post.
- [ ] Tests that assert you can see the comment each commentor left.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
Hi @MMhaidari 🤝 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
👍🏽 The PR follows the correct git flow ✔️ 
👍🏽 The PR has linters installed and it passes the checks ✔️ 
👍🏽 The PR has used the correct readme file template and all headers are present ✔️ 

## Required Changes ♻️
_Check the comments under the review._

- [ ] I saw that there are a lot of test cases that are failing, please make sure that you remove unnecessary ones and make sure that the ones you added are passing, I mean when we run test cases from your branch, everything should look green, kindly have a look at the following screenshot to learn about the test cases which are failing for now
Screenshot 2023-07-12 at 11 41 40


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by MMhaidari** [comment]
# What is added new 
- [ ]  Missing test cases added
- [ ] All tests are working fine

**Review by t-yanick — APPROVED** [review]
# STATUS: MILESTONE APPROVED :1st_place_medal: 

Hi @MMhaidari , 

### Great job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  All requested changes implemented

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**